### PR TITLE
Feature EPL-1525: How to Create a Computed Column

### PR DESCRIPTION
**Pull request by Tomas-Vesco**

### CODE REVIEW

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  * If it is a select sql expression, then all the tables in the from clause need to use an alias.
+  * You can refer to other columns of the table using their non-aliased name.
+
+This is an example of the total line quantity column shown above:
+
+```hql
+(SELECT sum(ol.qtyOrdered) FROM c_orderline ol WHERE ol.c_order_id=c_order_id)
+```
+
+What you can see is that the from clause uses an aliased table, the c_order_id at the end is the non-aliased column of the main table.
```
Comment (Leo-Echenique) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:52]: `c_order_id`

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+!!!Note
+    * A computed column can be used in the definition of a field for an Etendo column, just like every other column.
+    * A computed column that uses a String reference should have a length greater than zero. Because this is the number of characters which are going to be displayed in form view.
```
Comment (Leo-Echenique) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:56]: highlight 'String reference'

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+!!!Note
+    In case there are many computed columns in the same entity, they are evaluated all together when the first one is calculated.
+
+#### Limitations
+
+In order to make computed columns lazy, they are mapped in their entity as a `many-to-one` property (named `computedColumns`) linked to a virtual entity, is this virtual entity where the actual computed columns are located as hibernate formula properties.
+
+![](/assets/developer-guide/etendo-classic/how-to- guides/How_to_create_a_Computed_Column-6.png)
+
+This extra complexity is transparent when working with generated DAL Java classes, as seen before, `order.getDeliveryStatus()` populates the value of the computed column, though behind the scenes it is retrieving it from the virtual entity.
```
Comment (Leo-Echenique) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:115]: highlight DAL references

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  - How to
+  - SQL logic
+  - Etendo classic
+---
+
+# How to Create a Computed Column
+
+## Overview
+
+A computed column is a column of a table which is computed using an SQL expression and which does not exist explicitly in the database schema.
```
Comment (Leo-Echenique) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:13]: I think that the SQL reference should be highlighted, such as sqllogic field / whereclause. Everything that makes a references to an AD field that is commonly used.

**Review by Leo-Echenique — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+!!!Note
+  This code is not valid after 3.0MP027
+  FROM ORDER o WHERE o.deliveryStatus = 100
+
+doesn't work anymore starting from _MP27_ , its equivalent would be:
+
+    FROM ORDER o WHERE o._computedColumns.deliveryStatus = 100
+
+Starting from **3.0MP27** , computed columns cannot be used anymore in ` OBCriteria ` , if needed `OBQuery ` should be used instead.
```
Comment (valenvivaldi) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:134]: remove the " Starting from **3.0MP27** "

**Review by valeg-etendo — APPROVED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/.How_to_create_a_Computed_Column.md [hunk 6]
```diff
@@ -1,247 +0,0 @@
...
-
-![](/assets/developer-guide/etendo-classic/how-to-
-guides/How_to_create_a_Computed_Column-3.png){: .legacy-image-style}
-
-  
-or the form:
-
-  
-
-![](/assets/developer-guide/etendo-classic/how-to-
```
Comment (mapenglishservices) [inline docs/developer-guide/etendo-classic/how-to-guides/.How_to_create_a_Computed_Column.md:128]: Add ../../../ to all the links to make them relative

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+# How to Create a Computed Column
+
+## Overview
+
+A computed column is a column of a table which is computed using an _SQL_ expression and which does not exist explicitly in the database schema.
+
+Computed columns have these characteristics:
+
+  * Computed when a record is read from the database.
+  * Mot persisted, there is no real column in the database schema for a computed column.
```
Comment (Gremiger) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:18]:
```suggestion
  * Not persisted, there is no real column in the database schema for a computed column.
```

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+This extra complexity is transparent when working with generated _DAL_ Java classes, as seen before, `order.getDeliveryStatus()` populates the value of the computed column, though behind the scenes it is retrieving it from the virtual entity.
+
+But this model imposes also some limitations to be taken into account when manipulating computed columns in filters or sorting.
+
+!!!Note
+    When using computed columns in HQL to do filtering or sorting, they are evaluated. This needs to be done carefully (or even avoided if possible) because it might have performance implications.
+
+When working with HQL (i.e. `OBQuery`), computed columns cannot be directly accessed as they are not properties of the entity.
+
+So this valid query in versions prior to _MP27_:
```
Comment (RomanMagnoli) [inline docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:125]: :eyes:

File: docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+When working with HQL (i.e. `OBQuery`), computed columns cannot be directly accessed as they are not properties of the entity.
+
+So this valid query in versions prior to _MP27_ : 
+
+!!!Note
+  This code is not valid after 3.0MP027
+  FROM ORDER o WHERE o.deliveryStatus = 100
+
+doesn't work anymore starting from _MP27_ , its equivalent would be:
```
[the remark this answers is not in the document]
Comment (Gremiger) [reply docs/developer-guide/etendo-classic/how-to-guides/how-to-create-a-computed-column.md:131]: This was marked as resolved, but no chages were made

**Review by Leo-Echenique — APPROVED** [review]

**Review by RomanMagnoli — APPROVED** [review]

**Pull request merged**